src/Catalyst.Core.Modules.Dfs/DfsService.cs

```diff
@@ -0,0 +1,663 @@
+
+            var localPeer = new Peer
+            {
+                Id = self.Id,
+                PublicKey = await (await KeyChainAsync()).GetPublicKeyAsync("self").ConfigureAwait(false),
+                ProtocolVersion = "ipfs/0.1.0"
+            };
+
+            var version = typeof(DfsService).GetTypeInfo().Assembly.GetName().Version;
+            localPeer.AgentVersion = $"net-ipfs/{version.Major}.{version.Minor}.{version.Revision}";
+            Log.Debug("Built local peer");
+            LocalPeer = localPeer;
+
+            //Swarm
+            Log.Debug("Building swarm service");
+
+            if (Options.Swarm.PrivateNetworkKey == null)
+            {
+                var path = Path.Combine(Options.Repository.Folder, "swarm.key");
+                if (File.Exists(path))
+                {
+                    using (var x = File.OpenText(path))
+                    {
+                        Options.Swarm.PrivateNetworkKey = new PreSharedKey();
+                        Options.Swarm.PrivateNetworkKey.Import(x);
+                    }
+                }
+            }
+
+            SwarmService.LocalPeer = LocalPeer;
+            SwarmService.LocalPeerKey =
+                Key.CreatePrivateKey(await KeyApi.GetPrivateKeyAsync("self").ConfigureAwait(false));
+            SwarmService.NetworkProtector = Options.Swarm.PrivateNetworkKey == null
+                ? null
+                : new Psk1Protector
+                {
+                    Key = Options.Swarm.PrivateNetworkKey
+                };
+
+            if (Options.Swarm.PrivateNetworkKey != null)
+            {
+                Log.Debug($"Private network {Options.Swarm.PrivateNetworkKey.Fingerprint().ToHexString()}");
+            }
+
+            Log.Debug("Building bitswap service");
+            Log.Debug("Built bitswap service");
+            BitSwapService.SwarmService = SwarmService;
+            BitSwapService.BlockService = BlockApi;
+
+            Log.Debug("Building PubSub service");
+            PubSubService.LocalPeer = LocalPeer;
+            PubSubService.Routers.Add(new FloodRouter
+            {
+                SwarmService = SwarmService
+            });
+            Log.Debug("Built PubSub service");
+        }
+
+        /// <summary>
+        ///     Starts the network services.
+        /// </summary>
+        /// <returns>
+        ///     A task that represents the asynchronous operation.
+        /// </returns>
+        /// <remarks>
+        ///     Starts the various IPFS and Lib.P2P network services.  This should
+        ///     be called after any configuration changes.
+        /// </remarks>
+        /// <exception cref="Exception">
+        ///     When the engine is already started.
+        /// </exception>
+        public async Task StartAsync()
+        {
+            _dfsState.IsStarted = true;
+            if (_stopTasks.Count > 0)
+            {
+                throw new Exception("IPFS engine is already started.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)